File: examples/e2e-travel-concierge/agent/mcpClient.ts
// Thin MCP client. The agent is a real MCP client talking to the real MCP server
// over Streamable HTTP. One fresh client per call keeps it simple against the
// stateless server.
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';
import { env } from '../config.js';

export interface McpToolResult {
  isError: boolean;
  text: string;
  structured?: unknown;
}

export async function callMcpTool(
  name: string,
  args: Record<string, unknown>,
): Promise<McpToolResult> {
  const transport = new StreamableHTTPClientTransport(new URL(env.mcpServerUrl));
  const client = new Client({ name: 'travel-concierge-agent', version: '0.1.0' });
  await client.connect(transport);
  try {
    const result = await client.callTool({ name, arguments: args });
    const content = Array.isArray(result.content) ? result.content : [];
    const text = content
      .filter((c): c is { type: 'text'; text: string } => c.type === 'text')
      .map((c) => c.text)
      .join('\n');
    return { isError: result.isError === true, text, structured: result.structuredContent };
  } finally {
    await client.close();
  }
}

File: examples/e2e-travel-concierge-py/agent/mcp_client.py
# Thin MCP client, Python port of ../e2e-travel-concierge/agent/mcpClient.ts.
# The agent is a real MCP client talking to the real MCP server over
# Streamable HTTP, via the official `mcp` package. One fresh client per call
# keeps it simple against the stateless server -- Flask is synchronous, so
# each call spins up its own asyncio event loop for the duration of one
# request/response round trip.

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, Dict, Optional

from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client

from config import env


@dataclass
class McpToolResult:
    is_error: bool
    text: str
    structured: Optional[Dict[str, Any]] = None


async def _call_mcp_tool_async(name: str, args: Dict[str, Any]) -> McpToolResult:
    async with streamablehttp_client(env.mcp_server_url) as (read, write, _get_session_id):
        async with ClientSession(read, write) as session:
            await session.initialize()
            result = await session.call_tool(name, args)
            text = "\n".join(c.text for c in result.content if c.type == "text")
            return McpToolResult(
                is_error=bool(result.isError),
                text=text,
                structured=result.structuredContent,
            )


def call_mcp_tool(name: str, args: Dict[str, Any]) -> McpToolResult:
    return asyncio.run(_call_mcp_tool_async(name, args))
